File: app/src/main/java/dev/favourdevlabs/cleanthes/data/db/VaultDao.java
package dev.favourdevlabs.cleanthes.data.db;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import dev.favourdevlabs.cleanthes.data.entities.VaultEntry;

import java.util.List;
import java.util.ArrayList;

public class VaultDao {

    private final DatabaseHelper dbHelper;

    public VaultDao(DatabaseHelper dbHelper) {
        this.dbHelper = dbHelper;
    }

    public long insert(VaultEntry entry) {
        SQLiteDatabase db = dbHelper.getWritableDatabase();
        ContentValues values = entryToContentValues(entry);
        return db.insert(DatabaseHelper.TABLE_VAULT_ENTRIES, null, values);
    }

    public int update(VaultEntry entry) {
        SQLiteDatabase db = dbHelper.getWritableDatabase();
        ContentValues values = entryToContentValues(entry);
        String whereClause = DatabaseHelper.COLUMN_ID + " = ?";
        String[] whereArgs = { String.valueOf(entry.getId()) };
        return db.update(DatabaseHelper.TABLE_VAULT_ENTRIES, values, whereClause, whereArgs);
    }

    public int deleteById(long id) {
        SQLiteDatabase db = dbHelper.getWritableDatabase();
        String whereClause = DatabaseHelper.COLUMN_ID + " = ?";
        String[] whereArgs = { String.valueOf(id) };
        return db.delete(DatabaseHelper.TABLE_VAULT_ENTRIES, whereClause, whereArgs);
    }

    public int deleteAll() {
        SQLiteDatabase db = dbHelper.getWritableDatabase();
        return db.delete(DatabaseHelper.TABLE_VAULT_ENTRIES, null, null);
    }

    public List<VaultEntry> getAllEntries() {
        SQLiteDatabase db = dbHelper.getReadableDatabase();

        String orderBy = DatabaseHelper.COLUMN_IS_FAVORITE + " DESC, "
                + DatabaseHelper.COLUMN_TITLE + " ASC";

        Cursor cursor = db.query(DatabaseHelper.TABLE_VAULT_ENTRIES, null, null, null, null, null, orderBy);
        return cursorToList(cursor);
    }

    public VaultEntry getEntryById(long id) {
        SQLiteDatabase db = dbHelper.getReadableDatabase();
        String selection = DatabaseHelper.COLUMN_ID + " = ?";
        String[] selectionArgs = { String.valueOf(id) };

        Cursor cursor = db.query(DatabaseHelper.TABLE_VAULT_ENTRIES, null, selection, selectionArgs, null, null, null);

        if (cursor == null)
            return null;
        try {
            if (cursor.moveToFirst()) {
                return cursorToEntry(cursor);
            }
            return null;
        } finally {
            cursor.close();
        }
    }

    public List<VaultEntry> searchEntries(String query) {
        SQLiteDatabase db = dbHelper.getReadableDatabase();

        String likeQuery = "%" + query + "%";

        String selection = DatabaseHelper.COLUMN_TITLE + " LIKE ? OR "
                + DatabaseHelper.COLUMN_USERNAME + " LIKE ?";

        String[] selectionArgs = { likeQuery, likeQuery };

        String orderBy = DatabaseHelper.COLUMN_IS_FAVORITE + " DESC, "
                + DatabaseHelper.COLUMN_TITLE + " ASC";

        Cursor cursor = db.query(DatabaseHelper.TABLE_VAULT_ENTRIES, null, selection, selectionArgs, null, null,
                orderBy);
        return cursorToList(cursor);
    }

    public List<VaultEntry> getEntriesByCategory(String category) {
        SQLiteDatabase db = dbHelper.getReadableDatabase();

        String selection = DatabaseHelper.COLUMN_CATEGORY + " = ? ";
        String[] selectionArgs = { category };
        String orderBy = DatabaseHelper.COLUMN_IS_FAVORITE + " DESC, "
                + DatabaseHelper.COLUMN_TITLE + " ASC";

        Cursor cursor = db.query(DatabaseHelper.TABLE_VAULT_ENTRIES, null, selection, selectionArgs, null, null,
                orderBy);
        return cursorToList(cursor);
    }

    public List<VaultEntry> getFavoriteEntries() {
        SQLiteDatabase db = dbHelper.getReadableDatabase();

        String selection = DatabaseHelper.COLUMN_IS_FAVORITE + " = ?";
        String[] selectionArgs = { "1" };
        String orderBy = DatabaseHelper.COLUMN_TITLE + " ASC";

        Cursor cursor = db.query(DatabaseHelper.TABLE_VAULT_ENTRIES, null, selection, selectionArgs, null, null,
                orderBy);
        return cursorToList(cursor);
    }

    public int getEntryCount() {
        SQLiteDatabase db = dbHelper.getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT COUNT(*) FROM " + DatabaseHelper.TABLE_VAULT_ENTRIES, null);

        try {
            if (cursor.moveToFirst()) {
                return cursor.getInt(0);
            }
            return 0;
        } finally {
            cursor.close();
        }
    }

    public List<String> getAllCategories() {
        SQLiteDatabase db = dbHelper.getReadableDatabase();
        List<String> categories = new ArrayList<>();

        Cursor cursor = db.rawQuery("SELECT DISTINCT " + DatabaseHelper.COLUMN_CATEGORY + " FROM "
                + DatabaseHelper.TABLE_VAULT_ENTRIES + " ORDER BY " + DatabaseHelper.COLUMN_CATEGORY + " ASC", null);

        try {
            if (cursor.moveToFirst()) {
                do {
                    categories.add(cursor.getString(0));
                } while (cursor.moveToNext());
            }
        } finally {
            cursor.close();
        }
        return categories;
    }

    private ContentValues entryToContentValues(VaultEntry entry) {
        ContentValues values = new ContentValues();
        values.put(DatabaseHelper.COLUMN_TITLE, entry.getTitle());
        values.put(DatabaseHelper.COLUMN_USERNAME, entry.getUsername());
        values.put(DatabaseHelper.COLUMN_ENCRYPTED_PASSWORD, entry.getEncryptedPassword());
        values.put(DatabaseHelper.COLUMN_WEBSITE, entry.getWebsite());
        values.put(DatabaseHelper.COLUMN_CATEGORY, entry.getCategory());
        values.put(DatabaseHelper.COLUMN_NOTES, entry.getNotes());
        values.put(DatabaseHelper.COLUMN_CREATED_AT, entry.getCreatedAt());
        values.put(DatabaseHelper.COLUMN_UPDATED_AT, entry.getUpdatedAt());
        values.put(DatabaseHelper.COLUMN_IS_FAVORITE, entry.isFavorite() ? 1 : 0);
        return values;
    }

    private VaultEntry cursorToEntry(Cursor cursor) {
        VaultEntry entry = new VaultEntry();
        entry.setId(cursor.getLong(DatabaseHelper.IDX_ID));
        entry.setTitle(cursor.getString(DatabaseHelper.IDX_TITLE));
        entry.setUsername(cursor.getString(DatabaseHelper.IDX_USERNAME));
        entry.setEncryptedPassword(cursor.getString(DatabaseHelper.IDX_ENCRYPTED_PASSWORD));
        entry.setWebsite(cursor.getString(DatabaseHelper.IDX_WEBSITE));
        entry.setCategory(cursor.getString(DatabaseHelper.IDX_CATEGORY));
        entry.setNotes(cursor.getString(DatabaseHelper.IDX_NOTES));
        entry.setCreatedAt(cursor.getLong(DatabaseHelper.IDX_CREATED_AT));
        entry.setUpdatedAt(cursor.getLong(DatabaseHelper.IDX_UPDATED_AT));
        entry.setFavorite(cursor.getInt(DatabaseHelper.IDX_IS_FAVORITE) == 1);
        return entry;
    }

    private List<VaultEntry> cursorToList(Cursor cursor) {
        List<VaultEntry> entries = new ArrayList<>();
        if (cursor == null)
            return entries;

        try {
            if (cursor.moveToFirst()) {
                do {
                    entries.add(cursorToEntry(cursor));
                } while (cursor.moveToNext());
            }
        } finally {
            cursor.close();
        }
        return entries;
    }
}

File: app/src/main/java/dev/favourdevlabs/cleanthes/data/db/VaultDao.kt
package dev.favourdevlabs.cleanthes.data.db

import android.content.ContentValues
import android.database.Cursor
import dev.favourdevlabs.cleanthes.data.entities.VaultEntry

class VaultDao(private val dbHelper:DatabaseHelper)
{

    fun insert(entry: VaultEntry): Long =
        dbHelper.writableDatabase.insert(
            DatabaseHelper.TABLE_VAULT_ENTRIES, null, toContentValues(entry)
        )

    fun update(entry: VaultEntry): Int =
        dbHelper.writableDatabase.update(
            DatabaseHelper.TABLE_VAULT_ENTRIES,
            toContentValues(entry),
            "${DatabaseHelper.COLUMN_ID} = ?",
            arrayOf(entry.id.toString())
        )

    fun deleteById(id: Long): Int =
        dbHelper.writableDatabase.delete(
            DatabaseHelper.TABLE_VAULT_ENTRIES,
            "${DatabaseHelper.COLUMN_ID} = ?",
            arrayOf(id.toString())
        )

    fun deleteAll(): Int =
        dbHelper.writableDatabase.delete(DatabaseHelper.TABLE_VAULT_ENTRIES, null, null)

    fun getAllEntries(): List<VaultEntry> {
        val order = "${DatabaseHelper.COLUMN_IS_FAVORITE} DESC, ${DatabaseHelper.COLUMN_TITLE} ASC"
        return cursorToList(
            dbHelper.readableDatabase.query(
                DatabaseHelper.TABLE_VAULT_ENTRIES, null, null, null, null, null, order
            )
        )
    }

    fun getEntryById(id: Long): VaultEntry? {
        val cursor = dbHelper.readableDatabase.query(
            DatabaseHelper.TABLE_VAULT_ENTRIES, null,
            "${DatabaseHelper.COLUMN_ID} = ?",
            arrayOf(id.toString()),
            null, null, null
        )
        return cursor.use { if (it.moveToFirst()) fromCursor(it) else null }
    }

    fun searchEntries(query: String): List<VaultEntry> {
        val like = "%$query%"
        val where = "${DatabaseHelper.COLUMN_TITLE} LIKE ? OR ${DatabaseHelper.COLUMN_USERNAME} LIKE ?"
        val order = "${DatabaseHelper.COLUMN_IS_FAVORITE} DESC, ${DatabaseHelper.COLUMN_TITLE} ASC"
        return cursorToList(
            dbHelper.readableDatabase.query(
                DatabaseHelper.TABLE_VAULT_ENTRIES, null,
                where, arrayOf(like, like), null, null, order
            )
        )
    }

    fun getEntriesByCategory(category: String): List<VaultEntry> {
        val order = "${DatabaseHelper.COLUMN_IS_FAVORITE} DESC, ${DatabaseHelper.COLUMN_TITLE} ASC"
        return cursorToList(
            dbHelper.readableDatabase.query(
                DatabaseHelper.TABLE_VAULT_ENTRIES, null,
                "${DatabaseHelper.COLUMN_CATEGORY} = ?",
                arrayOf(category), null, null, order
            )
        )
    }

    fun getFavoriteEntries(): List<VaultEntry> =
        cursorToList(
            dbHelper.readableDatabase.query(
                DatabaseHelper.TABLE_VAULT_ENTRIES, null,
                "${DatabaseHelper.COLUMN_IS_FAVORITE} = ?",
                arrayOf("1"), null, null,
                "${DatabaseHelper.COLUMN_TITLE} ASC"
            )
        )

    fun getEntryCount(): Int {
        val c = dbHelper.readableDatabase
            .rawQuery("SELECT COUNT(*) FROM ${DatabaseHelper.TABLE_VAULT_ENTRIES}", null)
        return c.use { if (it.moveToFirst()) it.getInt(0) else 0 }
    }

    fun getAllCategories(): List<String> {
        val list = mutableListOf<String>()
        val c = dbHelper.readableDatabase.rawQuery(
            "SELECT DISTINCT ${DatabaseHelper.COLUMN_CATEGORY} " +
            "FROM ${DatabaseHelper.TABLE_VAULT_ENTRIES} " +
            "ORDER BY ${DatabaseHelper.COLUMN_CATEGORY} ASC", null
        )
        c.use { if (it.moveToFirst()) do { list.add(it.getString(0)) } while (it.moveToNext()) }
        return list
    }

    // -------------------------------------------------------------------------

    private fun toContentValues(e: VaultEntry) = ContentValues().apply {
        put(DatabaseHelper.COLUMN_TITLE,              e.title)
        put(DatabaseHelper.COLUMN_USERNAME,           e.username)
        put(DatabaseHelper.COLUMN_ENCRYPTED_PASSWORD, e.encryptedPassword)
        put(DatabaseHelper.COLUMN_WEBSITE,            e.website)
        put(DatabaseHelper.COLUMN_CATEGORY,           e.category)
        put(DatabaseHelper.COLUMN_NOTES,              e.notes)
        put(DatabaseHelper.COLUMN_CREATED_AT,         e.createdAt)
        put(DatabaseHelper.COLUMN_UPDATED_AT,         e.updatedAt)
        put(DatabaseHelper.COLUMN_IS_FAVORITE,        if (e.isFavorite) 1 else 0)
        put(DatabaseHelper.COLUMN_TOTP_SECRET,        e.totpSecret)
        put(DatabaseHelper.COLUMN_TOTP_ISSUER,        e.totpIssuer)
        put(DatabaseHelper.COLUMN_TOTP_DIGITS,        e.totpDigits)
        put(DatabaseHelper.COLUMN_TOTP_PERIOD,        e.totpPeriod)
        put(DatabaseHelper.COLUMN_TOTP_ALGORITHM,     e.totpAlgorithm)
    }

    private fun fromCursor(c: Cursor) = VaultEntry(
        id                = c.getLong(DatabaseHelper.IDX_ID),
        title             = c.getString(DatabaseHelper.IDX_TITLE),
        username          = c.getString(DatabaseHelper.IDX_USERNAME),
        encryptedPassword = c.getString(DatabaseHelper.IDX_ENCRYPTED_PASSWORD),
        website           = c.getString(DatabaseHelper.IDX_WEBSITE),
        category          = c.getString(DatabaseHelper.IDX_CATEGORY),
        notes             = c.getString(DatabaseHelper.IDX_NOTES),
        createdAt         = c.getLong(DatabaseHelper.IDX_CREATED_AT),
        updatedAt         = c.getLong(DatabaseHelper.IDX_UPDATED_AT),
        isFavorite        = c.getInt(DatabaseHelper.IDX_IS_FAVORITE) == 1,
        totpSecret        = if (c.isNull(DatabaseHelper.IDX_TOTP_SECRET)) null
                            else c.getString(DatabaseHelper.IDX_TOTP_SECRET),
        totpIssuer        = if (c.isNull(DatabaseHelper.IDX_TOTP_ISSUER)) null
                            else c.getString(DatabaseHelper.IDX_TOTP_ISSUER),
        totpDigits        = if (c.isNull(DatabaseHelper.IDX_TOTP_DIGITS)) 6
                            else c.getInt(DatabaseHelper.IDX_TOTP_DIGITS),
        totpPeriod        = if (c.isNull(DatabaseHelper.IDX_TOTP_PERIOD)) 30
                            else c.getInt(DatabaseHelper.IDX_TOTP_PERIOD),
        totpAlgorithm     = if (c.isNull(DatabaseHelper.IDX_TOTP_ALGORITHM)) "SHA1"
                            else c.getString(DatabaseHelper.IDX_TOTP_ALGORITHM)
    )

    private fun cursorToList(c: Cursor): List<VaultEntry> {
        val list = mutableListOf<VaultEntry>()
        c.use { if (it.moveToFirst()) do { list.add(fromCursor(it)) } while (it.moveToNext()) }
        return list
    }
}
